ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -184,8 +191,16 @@ addBlock cfg blk m
     blocks' :: Map (HeaderHash blk) blk
     blocks' = Map.insert (Block.blockHash blk) blk (blocks m)
 
-    candidates :: [(Chain blk, ExtLedgerState blk)]
-    candidates = mapMaybe (validate cfg (initLedger m)) $ chains blocks'
+    invalidBlocks' :: Set (Point blk)
+    candidates     :: [(Chain blk, ExtLedgerState blk)]
+    (invalidBlocks', candidates) =
+      foldMap (classify . validate cfg (initLedger m)) $ chains blocks'
+
+    classify :: ValidationResult blk
+             -> (Set (Point blk), [(Chain blk, ExtLedgerState blk)])
+    classify (ValidChain chain ledger) = (mempty, [(chain, ledger)])
+    classify (InvalidChain _ invalid chain ledger) =
+      (Set.fromList (NE.toList invalid), [(chain, ledger)])
```

Comment:
If the memory use / incremental calculation is important here then it may be better to have just one list result of (chain, extledger state, invalid blocks), and then accumulate the invalid blocks as we process the list of candidates.

Reply:
Ohhhhhh, just noticed this is the model so we don't care about the efficiency, just clarity. Ok, move on :-)